test/extensions/filters/network/dubbo_proxy/route_matcher_test.cc

```diff
@@ -914,6 +914,298 @@ name: config
+
+    EXPECT_EQ("k2", mmc[1]->name());
+    EXPECT_EQ(hv2, mmc[1]->value());
+  }
+
+  // Match with no metadata.
+  {
+    MessageMetadata metadata;
+    metadata.setInvocationInfo(invo);
+    invo->setServiceName("org.apache.dubbo.demo.DemoService");
+    invo->setMethodName("method2");
+
+    setDefaultInvocationCallback(*invo);
+
+    NiceMock<Server::Configuration::MockServerFactoryContext> context;
+    SingleRouteMatcherImpl matcher(config, context);
+    RouteConstSharedPtr route = matcher.route(metadata, 0);
+    EXPECT_NE(nullptr, route);
+    EXPECT_NE(nullptr, route->routeEntry());
+    EXPECT_EQ(nullptr, route->routeEntry()->metadataMatchCriteria());
+  }
+}
+
+TEST(DubboRouteMatcherTest, WeightedClusterMetadataMatch) {
+  const std::string yaml = R"EOF(
+name: config
+interface: org.apache.dubbo.demo.DemoService
+routes:
+  - match:
+      method:
+        name:
+          exact: "method1"
+    route:
+      weighted_clusters:
+        clusters:
+          - name: cluster1
+            weight: 2000
+            metadata_match:
+              filter_metadata:
+                envoy.lb:
+                  k1: v1
+                  k2: v2
+          - name: cluster2
+            weight: 3000
+            metadata_match:
+              filter_metadata:
+                not.envoy.lb:
+                  k1: v1
+                  k2: v2
+          - name: cluster3
+            weight: 5000
+            metadata_match:
+              filter_metadata:
+                envoy.lb:
+                  k3: v3
+)EOF";
+
+  envoy::extensions::filters::network::dubbo_proxy::v3::RouteConfiguration config =
+      parseRouteConfigurationFromV2Yaml(yaml);
+  auto invo = std::make_shared<RpcInvocationImpl>();
+  MessageMetadata metadata;
+  metadata.setInvocationInfo(invo);
+  invo->setServiceName("org.apache.dubbo.demo.DemoService");
+  invo->setMethodName("method1");
+
+  setDefaultInvocationCallback(*invo);
+  NiceMock<Server::Configuration::MockServerFactoryContext> context;
+  SingleRouteMatcherImpl matcher(config, context);
+
+  ProtobufWkt::Value v1, v2, v3;
+  v1.set_string_value("v1");
+  v2.set_string_value("v2");
+  v3.set_string_value("v3");
+  HashedValue hv1(v1), hv2(v2), hv3(v3);
+
+  // Weight cluster match with multiple metadata entries.
+  {
+
+    RouteConstSharedPtr route = matcher.route(metadata, 0);
+    EXPECT_NE(nullptr, route);
+    EXPECT_NE(nullptr, route->routeEntry());
+
+    const Envoy::Router::MetadataMatchCriteria* criteria =
+        route->routeEntry()->metadataMatchCriteria();
+    EXPECT_NE(nullptr, criteria);
+    const std::vector<Envoy::Router::MetadataMatchCriterionConstSharedPtr>& mmc =
+        criteria->metadataMatchCriteria();
+    EXPECT_EQ(2, mmc.size());
+
+    EXPECT_EQ("k1", mmc[0]->name());
+    EXPECT_EQ(hv1, mmc[0]->value());
+
+    EXPECT_EQ("k2", mmc[1]->name());
+    EXPECT_EQ(hv2, mmc[1]->value());
+  }
+
+  // Weighted cluster match with metadata key with "not.envoy.lb".
```

Comment:
May be just `The case where none 'envoy.lb' metadata key is used in the 'metadata_match'`